exchange/auction.go

```diff
@@ -60,111 +60,118 @@ func (a *auction) setRoundedPrices(priceGranularity openrtb_ext.PriceGranularity
-	expByImp := make(map[string]int64)
-	competitiveExclusion := false
-	var hbCacheID string
-	if len(bidCategory) > 0 {
-		// assert:  category of winning bids never duplicated
-		if rawUuid, err := uuid.NewV4(); err == nil {
-			hbCacheID = rawUuid.String()
-			competitiveExclusion = true
-		} else {
-			errs = append(errs, errors.New("failed to create custom cache key"))
-		}
-	}
 
-	// Grab the imp TTLs
-	for _, imp := range bidRequest.Imp {
-		expByImp[imp.ID] = imp.Exp
-	}
-	for _, topBidsPerImp := range a.winningBidsByBidder {
-		for _, topBidPerBidder := range topBidsPerImp {
-			impID := topBidPerBidder.bid.ImpID
-			var customCacheKey string
-			var catDur string
-			useCustomCacheKey := false
-			if competitiveExclusion && topBidPerBidder == a.winningBids[impID] {
-				// set custom cache key for winning bid when competitive exclusion applies
-				catDur = bidCategory[topBidPerBidder.bid.ID]
-				if len(catDur) > 0 {
-					customCacheKey = fmt.Sprintf("%s_%s", catDur, hbCacheID)
-					useCustomCacheKey = true
-				}
+	if (bids || vast) && (includeBidderKeys || includeWinners) {
+		expectNumBids := valOrZero(bids, len(a.roundedPrices))
+		expectNumVast := valOrZero(vast, len(a.roundedPrices))
+		bidIndices := make(map[int]*openrtb.Bid, expectNumBids)
+		vastIndices := make(map[int]*openrtb.Bid, expectNumVast)
+		toCache := make([]prebid_cache_client.Cacheable, 0, expectNumBids+expectNumVast)
+		expByImp := make(map[string]int64)
+		competitiveExclusion := false
+		var hbCacheID string
+		if len(bidCategory) > 0 {
+			// assert:  category of winning bids never duplicated
+			if rawUuid, err := uuid.NewV4(); err == nil {
+				hbCacheID = rawUuid.String()
+				competitiveExclusion = true
+			} else {
+				errs = append(errs, errors.New("failed to create custom cache key"))
 			}
-			if bids {
-				if jsonBytes, err := json.Marshal(topBidPerBidder.bid); err == nil {
-					if useCustomCacheKey {
-						// not allowed if bids is true; log error and cache normally
-						errs = append(errs, errors.New("cannot use custom cache key for non-vast bids"))
+		}
+
+		// Grab the imp TTLs
+		for _, imp := range bidRequest.Imp {
+			expByImp[imp.ID] = imp.Exp
+		}
+		for ImpName, topBidsPerImp := range a.winningBidsByBidder {
+			//	winningBidsByBidder map[ImpName]map[openrtb_ext.BidderName]*pbsOrtbBid
+			for bidderName, topBidPerBidder := range topBidsPerImp {
+				impID := topBidPerBidder.bid.ImpID
+				isOverallWinner := a.winningBids[impID] == topBidPerBidder
+				//  PBS-330 add a condition so that we don't cache what is not an overallWinner
+				//if targData.includeBidderKeys
+				//if targData.includeWinners && overallWinner
+				if includeBidderKeys || (includeWinners && isOverallWinner) { //overallWinner := a.winningBids[impId]
+					var customCacheKey string
+					var catDur string
+					useCustomCacheKey := false
+					if competitiveExclusion && topBidPerBidder == a.winningBids[impID] {
```

Comment:
use **isOverallWinner** here instead of **topBidPerBidder == a.winningBids[impID]**

Reply:
done